src/java/org/apache/cassandra/cql3/selection/ColumnTimestamps.java

```diff
@@ -0,0 +1,381 @@
+        {
+            // If this method is called it means that it is an element selection on a frozen collection/UDT,
+            // so we can safely return this Timestamps as all the elements also share that timestamp
+            return this;
+        }
+
+        @Override
+        public ColumnTimestamps max()
+        {
+            return this;
+        }
+
+        @Override
+        public ColumnTimestamps slice(Range<Integer> range)
+        {
+            return range.isEmpty() ? NO_TIMESTAMP : this;
+        }
+
+        @Override
+        public ByteBuffer toByteBuffer(ProtocolVersion protocolVersion)
+        {
+            return timestamp == type.defaultValue() ? null : type.toByteBuffer(timestamp);
+        }
+
+        @Override
+        public String toString()
+        {
+            return type + ": " + timestamp;
+        }
+    }
+
+    /**
+     * A {@link ColumnTimestamps} that can contain multiple timestamps (for unfrozen collections or UDTs).
+     */
+    private static final class MultipleTimestamps extends ColumnTimestamps
+    {
+        private final List<Long> timestamps;
+
+        public MultipleTimestamps(TimestampsType type, int initialCapacity)
+        {
+            this(type, new ArrayList<>(initialCapacity));
+        }
+
+        public MultipleTimestamps(TimestampsType type, List<Long> timestamps)
+        {
+            super(type);
+            this.timestamps = timestamps;
+        }
+
+        @Override
+        public void addNoTimestamp()
+        {
+            timestamps.add(type.defaultValue());
+        }
+
+        @Override
+        public void addTimestampFrom(Cell<?> cell, int nowInSecond)
+        {
+            timestamps.add(type.getTimestamp(cell, nowInSecond));
+        }
+
+        @Override
+        public ColumnTimestamps get(int index)
+        {
+            return timestamps.isEmpty()
+                   ? NO_TIMESTAMP
+                   : new SingleTimestamps(type, timestamps.get(index));
+        }
+
+        @Override
+        public ColumnTimestamps max()
+        {
+            return timestamps.isEmpty()
+                   ? NO_TIMESTAMP
+                   : new SingleTimestamps(type, Collections.max(timestamps));
+        }
+
+        @Override
+        public ColumnTimestamps slice(Range<Integer> range)
+        {
+            if (range.isEmpty())
+                return NO_TIMESTAMP;
+
+            int from = !range.hasLowerBound() ? 0
+                                              : range.lowerBoundType() == BoundType.CLOSED ? range.lowerEndpoint()
+                                                                                           : range.lowerEndpoint() + 1;
+            int to = !range.hasUpperBound() ? timestamps.size()
+                                            : range.upperBoundType() == BoundType.CLOSED ? range.upperEndpoint() + 1
+                                                                                         : range.upperEndpoint();
```

Comment:
nit: just a different style to avoid the nested ternary operators. 
Can you also add comments to explain calculations for `from` and `to`?

```suggestion
            int from = 0;
            int to = timestamps.size();
            if (range.hasLowerBound())
            {
                from = range.lowerBoundType() == BoundType.CLOSED
                       ? range.lowerEndpoint()
                       : range.lowerEndpoint() + 1;
            }
            if (range.hasUpperBound())
            {
                to = range.upperBoundType() == BoundType.CLOSED
                     ? range.upperEndpoint() + 1
                     : range.upperEndpoint();
            }
```

Reply:
Changed to the proposed format and added some comments about the calculations.